awscli/compat.py

```
@@ -120,9 +122,20 @@ def get_binary_stdout():
 
 
 def _get_text_writer(stream, errors):
+    set_preferred_output_encoding(stream)
     return stream
 
 
+def set_preferred_output_encoding(stream):
+    """
+    If the user specified AWS_CLI_OUTPUT_ENCODING, use that encoding
+    for the output stream. This is useful for Windows users since
+    PyInstaller 6 no longer honors PYTHONUTF8.
+    """
+    if OUTPUT_ENCODING_ENV_VAR in os.environ:
+        stream.reconfigure(encoding=os.environ[OUTPUT_ENCODING_ENV_VAR])
```

I think we should raise a more helpful exception here if an invalid encoding is configured.

---

Added via https://github.com/aws/aws-cli/pull/9319/commits/fdeac1e8079c527e42d03ca54893272c31242f06

```
(.venv312) ➜  aws-cli git:(shovlia/output-formatting) ✗ AWS_CLI_OUTPUT_ENCODING=invalid aws dynamodb get-item --table-name shovlia-manual --key '{"id": {"S": "encoding"}}'

Unknown codec `invalid` specified for AWS_CLI_OUTPUT_ENCODING.
```

I had initially tried to do it during `set_preferred_output_encoding`, but that can be called during error handling as we're preparing to write the output for some other exception. I wanted to keep writing out the original exception (without the invalid encoding), rather than now fail for this reason.

So `set_preferred_output_encoding` only logs an invalid value, but we will fail earlier from `clidriver` via `validate_preferred_output_encoding`.